File: packages/runpane-py/src/runpane/download.py
from __future__ import annotations

import hashlib
import os
import shutil
import tempfile
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Optional

from .releases import ResolvedRelease, artifact_file_name


@dataclass
class DownloadedArtifact:
    path: str
    file_name: str
    used_fallback: bool


def download_artifact(resolved: ResolvedRelease, download_dir: Optional[str], verbose: bool) -> DownloadedArtifact:
    target_dir = download_dir or os.path.join(tempfile.gettempdir(), f"runpane-{int(time.time() * 1000)}")
    os.makedirs(target_dir, exist_ok=True)

    file_name = artifact_file_name(resolved.artifact["name"])
    target_path = os.path.join(target_dir, file_name)
    used_fallback = False

    try:
        download_to_file(resolved.preferred_download_url, target_path, verbose)
    except Exception as error:
        used_fallback = True
        print(f"runpane: website download route failed; falling back to GitHub release asset. {error}")
        download_to_file(resolved.fallback_download_url, target_path, verbose)

    verify_checksum_if_available(resolved, target_path, file_name)
    return DownloadedArtifact(path=target_path, file_name=file_name, used_fallback=used_fallback)


def download_to_file(url: str, target_path: str, verbose: bool) -> None:
    if verbose:
        print(f"Downloading {url}")
    req = urllib.request.Request(url, headers={"User-Agent": "runpane-installer"})
    with urllib.request.urlopen(req, timeout=120) as response:
        if getattr(response, "status", 200) >= 400:
            raise RuntimeError(f"{response.status} {response.reason}")
        with open(target_path, "wb") as target:
            shutil.copyfileobj(response, target, length=1024 * 1024)


def verify_checksum_if_available(resolved: ResolvedRelease, artifact_path: str, file_name: str) -> None:
    try:
        req = urllib.request.Request(resolved.checksum_url, headers={"User-Agent": "runpane-installer"})
        with urllib.request.urlopen(req, timeout=30) as response:
            checksums = response.read().decode("utf-8")
    except (urllib.error.URLError, TimeoutError, OSError) as error:
        print(f"runpane: could not verify checksum for {file_name}. {error}")
        return

    expected = parse_checksum(checksums, file_name)
    if not expected:
        return

    digest = hashlib.sha256()
    with open(artifact_path, "rb") as source:
        for chunk in iter(lambda: source.read(1024 * 1024), b""):
            digest.update(chunk)

    actual = digest.hexdigest()
    if actual.lower() != expected.lower():
        raise RuntimeError(f"Checksum mismatch for {file_name}. Expected {expected}, got {actual}.")


def parse_checksum(checksums: str, file_name: str) -> Optional[str]:
    for line in checksums.splitlines():
        stripped = line.strip()
        if not stripped.endswith(file_name):
            continue
        digest = stripped.split()[0]
        if len(digest) == 64 and all(char in "0123456789abcdefABCDEF" for char in digest):
            return digest
    return None

File: packages/runpane/src/download.ts
import crypto from 'crypto';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { Readable } from 'stream';
import { pipeline } from 'stream/promises';
import type { ReadableStream as NodeReadableStream } from 'stream/web';
import { artifactFileName, type ResolvedRelease } from './releases';

export interface DownloadedArtifact {
  path: string;
  fileName: string;
  usedFallback: boolean;
}

export async function downloadArtifact(
  resolved: ResolvedRelease,
  downloadDir?: string,
  verbose = false,
  onFallbackUsed?: (error: unknown) => Promise<void> | void
): Promise<DownloadedArtifact> {
  const targetDir = downloadDir ?? path.join(os.tmpdir(), `runpane-${Date.now()}`);
  fs.mkdirSync(targetDir, { recursive: true });

  const fileName = artifactFileName(resolved.artifact.name);
  const targetPath = path.join(targetDir, fileName);

  let usedFallback = false;
  try {
    await downloadToFile(resolved.preferredDownloadUrl, targetPath, verbose);
  } catch (error) {
    usedFallback = true;
    console.warn(`runpane: website download route failed; falling back to GitHub release asset. ${formatError(error)}`);
    try {
      await onFallbackUsed?.(error);
    } catch {
      // Fallback telemetry must not affect download behavior.
    }
    await downloadToFile(resolved.fallbackDownloadUrl, targetPath, verbose);
  }

  await verifyChecksumIfAvailable(resolved, targetPath, fileName);
  return { path: targetPath, fileName, usedFallback };
}

async function downloadToFile(url: string, targetPath: string, verbose: boolean): Promise<void> {
  if (verbose) {
    console.log(`Downloading ${url}`);
  }

  const response = await fetch(url, {
    redirect: 'follow',
    headers: { 'user-agent': 'runpane-installer' }
  });

  if (!response.ok || !response.body) {
    throw new Error(`${response.status} ${response.statusText}`);
  }

  await pipeline(
    Readable.fromWeb(response.body as unknown as NodeReadableStream<Uint8Array>),
    fs.createWriteStream(targetPath)
  );
}

async function verifyChecksumIfAvailable(resolved: ResolvedRelease, artifactPath: string, fileName: string): Promise<void> {
  try {
    const response = await fetch(resolved.checksumUrl, {
      headers: { 'user-agent': 'runpane-installer' }
    });
    if (!response.ok) {
      return;
    }
    const checksums = await response.text();
    const expected = parseChecksum(checksums, fileName);
    if (!expected) {
      return;
    }
    const hash = crypto.createHash('sha256').update(fs.readFileSync(artifactPath)).digest('hex');
    if (hash.toLowerCase() !== expected.toLowerCase()) {
      throw new Error(`Checksum mismatch for ${fileName}. Expected ${expected}, got ${hash}.`);
    }
  } catch (error) {
    if (error instanceof Error && error.message.startsWith('Checksum mismatch')) {
      throw error;
    }
    console.warn(`runpane: could not verify checksum for ${fileName}. ${formatError(error)}`);
  }
}

function parseChecksum(checksums: string, fileName: string): string | undefined {
  for (const line of checksums.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed.endsWith(fileName)) continue;
    const [hash] = trimmed.split(/\s+/);
    if (/^[a-f0-9]{64}$/i.test(hash)) {
      return hash;
    }
  }
  return undefined;
}

function formatError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
